// @unibeautify/beautifier-php-cs-fixer/dist/src/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const unibeautify_1 = require("unibeautify");
const readPkgUp = require("read-pkg-up");
const tmp = require("tmp");
const fs = require("fs");
const path = require("path");
const os = require("os");
const { pkg } = readPkgUp.sync({ cwd: __dirname });
exports.beautifier = {
    name: "PHP-CS-Fixer",
    package: pkg,
    badges: [
        {
            description: "Build Status",
            url: "https://travis-ci.com/Unibeautify/beautifier-php-cs-fixer.svg?branch=master",
            href: "https://travis-ci.com/Unibeautify/beautifier-php-cs-fixer",
        },
        {
            description: "Test Coverage",
            url: "https://api.codeclimate.com/v1/badges/5412cdac960ca7eda78f/test_coverage",
            href: "https://codeclimate.com/github/Unibeautify/beautifier-php-cs-fixer/test_covera" +
                "ge",
        },
        {
            description: "Maintainability",
            url: "https://api.codeclimate.com/v1/badges/5412cdac960ca7eda78f/maintainability",
            href: "https://codeclimate.com/github/Unibeautify/beautifier-php-cs-fixer/maintainabi" +
                "lity",
        },
        {
            description: "Greenkeeper",
            url: "https://badges.greenkeeper.io/Unibeautify/beautifier-php-cs-fixer.svg",
            href: "https://greenkeeper.io/",
        },
    ],
    options: {
        PHP: true,
    },
    dependencies: [
        {
            type: unibeautify_1.DependencyType.Executable,
            name: "PHP-CS-Fixer",
            program: "php-cs-fixer",
            parseVersion: [/version (.*) by/, /PHP CS Fixer (\d+\.\d+\.\d+)/],
            homepageUrl: "https://cs.sensiolabs.org/",
            installationUrl: "https://cs.sensiolabs.org/#installation",
            bugsUrl: "https://github.com/FriendsOfPHP/PHP-CS-Fixer/issues",
            badges: [
                {
                    description: "Build Status (Travis CI)",
                    url: "https://travis-ci.org/FriendsOfPHP/PHP-CS-Fixer.svg",
                    href: "https://travis-ci.org/FriendsOfPHP/PHP-CS-Fixer",
                },
                {
                    description: "Code Coverage",
                    url: "https://coveralls.io/repos/FriendsOfPHP/PHP-CS-Fixer/badge.svg?service=github",
                    href: "https://coveralls.io/github/FriendsOfPHP/PHP-CS-Fixer",
                },
            ],
        },
    ],
    resolveConfig: ({ filePath, projectPath }) => {
        const configFiles = [".php_cs", ".php_cs.dist"];
        return findFile({
            finishPath: projectPath,
            startPath: filePath,
            fileNames: configFiles,
        })
            .then(configFile => ({ filePath: configFile }))
            .catch(err => {
            // tslint:disable-next-line no-console
            console.error(err);
            return Promise.resolve({});
        });
    },
    beautify({ text, dependencies, filePath, beautifierConfig, }) {
        const phpCsFixer = dependencies.get("PHP-CS-Fixer");
        const basePath = os.tmpdir();
        const config = beautifierConfig && beautifierConfig.filePath
            ? `--config=${beautifierConfig.filePath}`
            : "";
        // tslint:disable-next-line no-console
        console.error(`Using config: ${config}`);
        return tmpFile({ postfix: ".php" }).then(filePath => writeFile(filePath, text).then(() => phpCsFixer
            .run({
            args: relativizePaths(["fix", config, "--using-cache=no", filePath], basePath),
            options: {
                cwd: basePath,
            },
        })
            .then(({ exitCode, stderr }) => {
            if (exitCode) {
                return Promise.reject(stderr);
            }
            return readFile(filePath);
        })));
    },
};
function findFile({ finishPath = "/", startPath = finishPath, fileNames, }) {
    const filePaths = fileNames.map(fileName => path.join(startPath, fileName));
    return Promise.all(filePaths.map(doesFileExist))
        .then(exists => filePaths.filter((filePath, index) => exists[index]))
        .then(foundFilePaths => {
        if (foundFilePaths.length > 0) {
            return foundFilePaths[0];
        }
        if (startPath === finishPath) {
            return Promise.reject("No config file found");
        }
        const parentDir = path.resolve(startPath, "../");
        return findFile({ startPath: parentDir, finishPath, fileNames });
    });
}
function doesFileExist(filePath) {
    return new Promise(resolve => {
        fs.access(filePath, fs.constants.R_OK, error => resolve(!error));
    });
}
function tmpFile(options) {
    return new Promise((resolve, reject) => tmp.file(Object.assign({ prefix: "unibeautify-" }, options), (err, path, fd) => {
        if (err) {
            return reject(err);
        }
        return resolve(path);
    }));
}
function writeFile(filePath, contents) {
    return new Promise((resolve, reject) => {
        fs.writeFile(filePath, contents, error => {
            if (error) {
                return reject(error);
            }
            return resolve();
        });
    });
}
function readFile(filePath) {
    return new Promise((resolve, reject) => {
        fs.readFile(filePath, (error, data) => {
            if (error) {
                return reject(error);
            }
            return resolve(data.toString());
        });
    });
}
function relativizePaths(args, basePath) {
    return args.map(arg => {
        const isTmpFile = typeof arg === "string" &&
            !arg.includes(":") &&
            path.isAbsolute(arg) &&
            path.dirname(arg).startsWith(basePath);
        if (isTmpFile) {
            return path.relative(basePath, arg);
        }
        return arg;
    });
}
exports.default = exports.beautifier;
